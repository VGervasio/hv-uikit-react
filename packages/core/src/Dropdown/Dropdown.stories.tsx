import { Story, Meta } from "@storybook/react";
import Dropdown, { DropdownProps } from "./Dropdown";

export default {
  component: Dropdown,
  title: "Dropdown",
} as Meta;

const Template: Story<DropdownProps> = (args) => <Dropdown {...args} />;

export const Primary = Template.bind({});

Primary.args = {
  label: "Some Label",
  value: "someValue",
  options: [{ label: "Some Label", value: "someValue" }],
};
