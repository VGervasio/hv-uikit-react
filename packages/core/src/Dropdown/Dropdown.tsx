import styled from "@emotion/styled";

export interface DropdownOption {
  value: string;
  label: string;
}

export interface DropdownProps {
  value?: string;
  options: DropdownOption[];
  onChange?: (event: React.ChangeEvent<HTMLSelectElement>) => void;
}

const Dropdown: React.FC<DropdownProps> = ({ value, options, onChange }) => {
  const StyledDropdown = styled.div({});

  return (
    <StyledDropdown>
      <select value={value} onChange={onChange}>
        {options.map((option) => (
          <option key={option.value} value={option.value}>
            {option.label}
          </option>
        ))}
      </select>
    </StyledDropdown>
  );
};

export default Dropdown;
