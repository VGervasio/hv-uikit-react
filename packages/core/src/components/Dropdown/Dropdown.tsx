import styled from "@emotion/styled";
import { themeVars, themeUtils } from "theme";

export interface DropdownOption {
  value: string;
  label: string;
}

export interface DropdownProps {
  value?: string;
  options: DropdownOption[];
  onChange?: (event: React.ChangeEvent<HTMLSelectElement>) => void;
}

const StyledDropdown = styled("select")({
  border: `1px solid ${themeVars.colors.atmo4}`,
  "&:hover": {
    border: `1px solid ${themeVars.colors.acce1}`,
  },
  height: themeVars.space.base,
  padding: `0 ${themeUtils.spacing(1)} 0 ${themeUtils.spacing(1)}`,
  "& select": {
    color: themeVars.colors.acce1,
  },
});

const Dropdown: React.FC<DropdownProps> = ({ value, options, onChange }) => {
  return (
    <StyledDropdown value={value} onChange={onChange}>
      {options?.map((option) => (
        <option key={option.value} value={option.value}>
          {option.label}
        </option>
      ))}
    </StyledDropdown>
  );
};

export default Dropdown;

if (process.env.NODE_ENV !== "production") {
  Dropdown.displayName = "Dropdown";
}
