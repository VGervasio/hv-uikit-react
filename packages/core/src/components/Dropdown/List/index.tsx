export { default as dropdownListClasses } from "./listClasses";
export * from "./listClasses";
export * from "./List";
