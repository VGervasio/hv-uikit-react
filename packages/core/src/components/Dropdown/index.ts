export { default as dropdownClasses } from "./dropdownClasses";
export * from "./dropdownClasses";
export * from "./Dropdown";
