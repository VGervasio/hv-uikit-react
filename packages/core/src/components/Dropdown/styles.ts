import { themeUtils, themeVars } from "theme";

const styles = {
  // border: `1px solid ${themeVars.colors.atmo4}` as string,
  // width: "100%",
  // borderRadius: 2,
  // position: "relative",
  // display: "inline-block",
  // oadd

  cursor: "pointer",
  userSelect: "none",
  position: "relative",
  width: "100%",
  background: themeVars.colors.atmo1,
  border: `1px solid ${themeVars.colors.atmo4}`,
  "&:hover": {
    border: `1px solid ${themeVars.colors.acce1}`,
  },
  borderRadius: "2px",
  display: "flex",
  alignItems: "center",
  height: themeUtils.spacing(4),
  padding: `0 ${themeUtils.spacing(1)} 0 ${themeUtils.spacing(1)}`,
  "& select": {
    color: themeVars.colors.acce1,
  },
} as unknown as TemplateStringsArray;

export default styles;
