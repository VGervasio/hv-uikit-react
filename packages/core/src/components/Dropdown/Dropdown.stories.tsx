import { Meta } from "@storybook/react";
import { Dropdown } from "./Dropdown";
import { Box } from "../Box";

export default {
  component: Dropdown,
  title: "Dropdown",
} as Meta;

export const Primary = () => {
  return (
    <Box sx={{ width: 200, height: 200, position: "relative" }}>
      <Dropdown
        value="value1"
        options={[
          { label: "Label 1", value: "value1" },
          { label: "Label 2", value: "value2" },
          { label: "Label 3", value: "value3" },
        ]}
      />
    </Box>
  );
};
