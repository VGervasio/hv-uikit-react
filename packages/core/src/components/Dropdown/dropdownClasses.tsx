import { getClasses } from "utils";
import { HvDropdownListClasses } from "./List/dropdownListClasses";

export type HvDropdownClasses = {
  root?: string;
  labelContainer?: string;
  label?: string;
  description?: string;
  error?: string;
  placeholder?: string;
  selectionDisabled?: string;
  dropdown?: string;
  arrow?: string;
  dropdownHeader?: string;
  dropdownHeaderInvalid?: string;
  dropdownHeaderOpen?: string;
} & HvDropdownListClasses;

const classKeys: string[] = [
  "root",
  "labelContainer",
  "label",
  "description",
  "error",
  "placeholder",
  "selectionDisabled",
  "dropdown",
  "arrow",
  "dropdownHeader",
  "dropdownHeaderInvalid",
  "dropdownHeaderOpen",
  "rootList",
  "dropdownListContainer",
  "searchContainer",
  "listBorderDown",
  "listContainer",
  "selectAllContainer",
  "selection",
  "selectAll",
];

const dropdownClasses = getClasses<HvDropdownClasses>(classKeys, "HvDropdown");

export default dropdownClasses;
