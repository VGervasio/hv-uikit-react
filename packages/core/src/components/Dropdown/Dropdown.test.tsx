import React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it } from "vitest";
import { HvDropdown } from "./Dropdown";

const Main = () => (
  <div style={{ width: 310 }}>
    <HvDropdown
      aria-label="Main sample"
      multiSelect
      showSearch
      values={[
        { label: "value 1" },
        { label: "value 2", selected: true },
        { label: "value 3" },
        { label: "value 4" },
      ]}
    />
  </div>
);

describe("Dropdown", () => {
  it("should be defined", () => {
    const { container } = render(<HvDropdown />);
    expect(container).toBeDefined();
  });

  it("should render correctly", () => {
    const { container } = render(<Main />);
    expect(container).toMatchSnapshot();
  });

  it("should open dropdown on click", async () => {
    const { getByRole } = render(
      <div style={{ width: 310 }}>
        <HvDropdown
          aria-label="Main sample"
          multiSelect
          showSearch
          values={[
            { label: "value 1" },
            { label: "value 2", selected: true },
            { label: "value 3" },
            { label: "value 4" },
          ]}
        />
      </div>
    );

    const DropdownHeader = getByRole("combobox");

    expect(DropdownHeader).toHaveAttribute("aria-expanded", "false");

    await userEvent.click(DropdownHeader);

    expect(DropdownHeader).toHaveAttribute("aria-expanded", "true");
  });

  it("should be disabled", async () => {
    const { container, getByRole } = render(
      <div style={{ width: 310 }}>
        <HvDropdown
          aria-label="Main sample"
          multiSelect
          showSearch
          disabled
          values={[
            { label: "value 1" },
            { label: "value 2", selected: true },
            { label: "value 3" },
            { label: "value 4" },
          ]}
        />
      </div>
    );
    expect(
      container.getElementsByClassName("HvBaseDropdown-rootDisabled")
    ).toHaveLength(1);

    const DropdownHeader = getByRole("combobox");

    expect(DropdownHeader).toHaveAttribute("aria-expanded", "false");

    expect(DropdownHeader).toHaveAttribute("aria-expanded", "false");
  });

  it("should be invalid", async () => {
    const { getByRole } = render(
      <div style={{ width: 310 }}>
        <HvDropdown
          aria-label="Main sample"
          multiSelect
          showSearch
          status="invalid"
          values={[
            { label: "value 1" },
            { label: "value 2", selected: true },
            { label: "value 3" },
            { label: "value 4" },
          ]}
        />
      </div>
    );

    const DropdownHeader = getByRole("combobox");

    expect(DropdownHeader).toHaveAttribute("aria-invalid", "true");
  });

  it("should select one", async () => {
    const { getByRole, getAllByRole } = render(
      <div style={{ width: 310 }}>
        <HvDropdown
          aria-label="Main sample"
          multiSelect
          showSearch
          values={[
            { label: "value 1" },
            { label: "value 2" },
            { label: "value 3" },
            { label: "value 4" },
          ]}
        />
      </div>
    );

    const DropdownHeader = getByRole("combobox");

    expect(DropdownHeader).toHaveAttribute("aria-expanded", "false");

    await userEvent.click(DropdownHeader);

    expect(DropdownHeader).toHaveAttribute("aria-expanded", "true");

    const options = getAllByRole("option");

    expect(options).toHaveLength(4);

    expect(options[0]).toHaveAttribute("aria-selected", "false");

    await userEvent.click(options[0]);

    expect(options[0]).toHaveAttribute("aria-selected", "true");
  });

  it("read Only", async () => {
    const { getByRole } = render(
      <div style={{ width: 310 }}>
        <HvDropdown
          aria-label="Main sample"
          multiSelect
          showSearch
          readOnly
          values={[
            { label: "value 1" },
            { label: "value 2" },
            { label: "value 3" },
            { label: "value 4" },
          ]}
        />
      </div>
    );

    const DropdownHeader = getByRole("combobox");

    expect(DropdownHeader).toHaveAttribute("aria-expanded", "false");

    await userEvent.click(DropdownHeader);

    expect(DropdownHeader).toHaveAttribute("aria-expanded", "false");
  });
});
