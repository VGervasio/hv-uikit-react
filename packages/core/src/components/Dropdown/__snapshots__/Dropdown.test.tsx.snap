// Vitest Snapshot v1

exports[`Dropdown > should render correctly 1`] = `
<div>
  <div
    style="width: 310px;"
  >
    <div
      class="HvDropdown-root css-13v24hy-StyledHvFormElement e1jemzt54 HvFormElement-root"
    >
      <div
        class="HvBaseDropdown-root HvDropdown-dropdown css-1nukton-StyledRoot e1xgfj1t9"
      >
        <div
          aria-expanded="false"
          aria-label="Main sample"
          class="e1jemzt53 HvBaseDropdown-anchor css-151d1mt-StyledAnchor-StyledDropdown e1xgfj1t8"
          role="combobox"
          tabindex="-1"
        >
          <div
            aria-label="Main sample"
            class="HvBaseDropdown-header HvDropdown-dropdownHeader css-z15in3-StyledHeaderRoot e1xgfj1t7"
            role="textbox"
            tabindex="0"
          >
            <div
              class="HvBaseDropdown-selection css-1n3tnd9-StyledSelection e1xgfj1t6"
            >
              <div
                class="HvDropdown-placeholder e1jemzt50 css-4kh4g2-getStyledComponent-StyledTypography e1te0rfk0"
                variant="body"
              >
                <b>
                  1
                </b>
                 / 4
              </div>
            </div>
            <div
              class="HvBaseDropdown-arrow HvDropdown-arrow css-tgcv4t-StyledDropDownXS e1xgfj1t3 css-du18qm"
              name="DropDownXS"
            >
              <svg
                focusable="false"
                height="12"
                viewBox="0 0 12 12"
                width="12"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  class="color0"
                  d="M5.99994 8.65 1.4 4.05l.7-.7L6 7.25l3.9-3.9.7.7z"
                  fill="var(--colors-acce1)"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
